src/Lean/Compiler/LCNF/JoinPoints.lean

```diff
@@ -5,11 +5,94 @@ Authors: Henrik Böving
 -/
 import Lean.Compiler.LCNF.CompilerM
 import Lean.Compiler.LCNF.PassManager
+import Lean.Compiler.LCNF.PullFunDecls
 
 namespace Lean.Compiler.LCNF
 
+-- TODO: These can be used in a much more general context
+partial def mapFVarM [Monad m] (f : FVarId → m FVarId) (e : Expr) : m Expr := do
+  match e with
+  | .proj typ idx struct => return .proj typ idx (←mapFVarM f struct)
+  | .app fn arg => return .app (←mapFVarM f fn) (←mapFVarM f arg)
+  | .fvar fvarId => return .fvar (←f fvarId)
+  | .lam arg ty body bi =>
+    return .lam arg (←mapFVarM f ty) (←mapFVarM f body) bi
+  | .forallE arg ty body bi =>
+    return .forallE arg (←mapFVarM f ty) (←mapFVarM f body) bi
+  | .letE var ty value body nonDep  =>
```

Comment:
The `.letE` is unreachable in LCNF.

Reply:
I thought it was reachable in type declarations? That's why one of the reasons I asked you whether I should walk type stuff as well.